src/engine/Service.cpp

```diff
@@ -504,3 +495,117 @@ std::optional<std::string> Service::idToValueForValuesClause(
+      (!rightOnly && static_cast<bool>(a) == static_cast<bool>(b))) {
+    return;
+  }
+
+  const auto& [service, sibling] = [&]() {
+    if (a) {
+      return std::tie(a, right);
+    } else {
+      AD_CORRECTNESS_CHECK(b);
+      return std::tie(b, left);
+    }
+  }();
+
+  auto addRuntimeInfo = [&](bool siblingUsed) {
+    std::string_view v = siblingUsed ? "yes"sv : "no"sv;
+    service->runtimeInfo().addDetail("optimized-with-sibling-result", v);
+    sibling->runtimeInfo().addDetail("used-to-optimize-service-sibling", v);
+  };
+
+  auto siblingResult = sibling->getResult(
+      false, requestLaziness ? ComputationMode::LAZY_IF_SUPPORTED
+                             : ComputationMode::FULLY_MATERIALIZED);
+
+  if (siblingResult->isFullyMaterialized()) {
+    sibling->precomputedResultBecauseSiblingOfService_ = siblingResult;
+    bool resultIsSmall = siblingResult->idTable().size() <=
+                         RuntimeParameters().get<"service-max-value-rows">();
+    if (resultIsSmall) {
+      service->siblingInfo_.emplace(
+          siblingResult, sibling->getExternallyVisibleVariableColumns(),
+          sibling->getCacheKey());
+    }
+    addRuntimeInfo(resultIsSmall);
+    return;
+  }
+
+  // Creates an idTable-Generator from partially materialized result data.
+  auto partialResultGenerator =
+      [](std::vector<Result::IdTableVocabPair> pairs,
+         cppcoro::generator<Result::IdTableVocabPair> prevGenerator,
+         cppcoro::generator<Result::IdTableVocabPair>::iterator it)
+      -> cppcoro::generator<Result::IdTableVocabPair> {
+    for (auto& pair : pairs) {
+      co_yield pair;
+    }
+    for (auto& pair : std::ranges::subrange{it, prevGenerator.end()}) {
+      co_yield pair;
+    }
+  };
+
+  size_t rows = 0;
+  std::vector<Result::IdTableVocabPair> resultPairs;
+  auto generator = std::move(siblingResult->idTables());
+  auto it = generator.begin();
+  for (; it != generator.end(); ++it) {
+    auto& pair = (*it);
+    rows += pair.idTable_.size();
+    resultPairs.push_back(std::move(pair));
+
+    if (rows > RuntimeParameters().get<"service-max-value-rows">()) {
+      // Stop precomputation as the siblingResult's size exceeds the threshold
+      // it is not useful for the service operation.
+      // Pass the partially materialized result to the sibling.
+      sibling->precomputedResultBecauseSiblingOfService_ =
+          std::make_shared<const Result>(
+              partialResultGenerator(std::move(resultPairs),
+                                     std::move(generator), std::move(it)),
+              siblingResult->sortedBy());
+      addRuntimeInfo(false);
+      return;
+    }
+  }
+
+  // The siblingResult has been fully materialized, so it can now be
+  // used in both sibling and service.
+  Result::IdTableVocabPair siblingPair(
+      IdTable{sibling->getResultWidth(),
+              sibling->getExecutionContext()->getAllocator()},
+      LocalVocab{});
+
+  for (auto& pair : resultPairs) {
```

Comment:
You know or can easily compute the total number of rows and the `reserve` the `idTable`.